usr/src/lib/smbclnt/libfknsmb/common/fksmb_sign_pkcs.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright 2017 Nexenta Systems, Inc.  All rights reserved.
+ */
+
+/*
+ * Helper functions for SMB signing using PKCS#11
+ *
+ * There are two implementations of these functions:
+ * This one (for user space) and another for kernel.
+ * See: uts/common/fs/smbclnt/netsmb/smb_sign_kcf.c
+ */
+
+#include <stdlib.h>
+#include <strings.h>
+#include <netsmb/smb_signing.h>
+#include <security/cryptoki.h>
+#include <security/pkcs11.h>
+
+/*
+ * SMB1 signing helpers:
+ * (getmech, init, update, final)
+ */
```

Comment:
I'm surprised you had to fake this file so much.  I'd have thought it would only have been necessary to remap PKCS11 to the in-kernel ops, leaving more of your mainline code the same for test in user space.  This was probably easier though, and I don't care enough about it to request a change.

Reply:
The interfaces are just different enough so we needed a thin adapter layer.
This is it.  The alternative would be a ton of ifdefs, which I'd rather not have.